README.md
# Basic Shopify API

![Tests](https://github.com/gnikyt/Basic-Shopify-API/workflows/Package%20Test/badge.svg?branch=master)
[![codecov](https://codecov.io/gh/gnikyt/Basic-Shopify-API/branch/master/graph/badge.svg?token=qqUuLItqJj)](https://codecov.io/gh/gnikyt/Basic-Shopify-API)
[![License](https://poser.pugx.org/gnikyt/basic-shopify-api/license)](https://packagist.org/packages/gnikyt/basic-shopify-api)

A simple, tested, API wrapper for Shopify using Guzzle.

It supports both the sync/async REST and GraphQL API provided by Shopify, basic rate limiting, and request retries.

It contains helpful methods for generating a installation URL, an authorize URL (offline and per-user), HMAC signature validation, call limits, and API requests.

It works with both OAuth and private API apps.

## Table of Contents
  * [Installation](#installation)
  * [Usage](#usage)
      * [Public API](#public-api)
        * [REST (sync)](#rest-sync)
        * [REST (async)](#rest-async)
        * [GraphQL (sync)](#graphql-sync)
        * [GraphQL (async)](#graphql-async)
        * [Getting access (offline)](#getting-access-offline)
        * [Getting access (per-user)](#getting-access-per-user)
        * [Verifying HMAC signature](#verifying-hmac-signature)
      * [Private API](#private-api)
        * [REST](#rest)
        * [GraphQL](#graphql)
      * [Making requests](#making-requests)
        * [REST](#rest-1)
            * [If sync is true (regular rest call)](#if-sync-is-true-regular-rest-call)
            * [If sync is false (restAsync call)](#if-sync-is-false-restasync-call)
            * [Overriding request type](#overriding-request-type)
            * [Passing additional request options](#passing-additional-request-options)
        * [GraphQL](#graphql-1)
      * [API Versioning](#api-versioning)
      * [Rate Limiting](#rate-limiting)
      * [page_info / pagination Support](#page_info--pagination-support)
      * [Isolated API calls](#isolated-api-calls)
      * [Retries](#retries)
      * [Errors](#errors)
      * [Middleware](#middleware)
      * [Storage](#storage)
  * [Documentation](#documentation)
  * [LICENSE](#license)

## Installation

The recommended way to install is [through composer](http://packagist.org).

    composer require gnikyt/basic-shopify-api

## Usage

### Public API

This assumes you properly have your app setup in the partner's dashboard with the correct keys and redirect URIs.

#### REST (sync)

For REST calls, the shop domain and access token are required.

```php
use Gnikyt\BasicShopifyAPI\BasicShopifyAPI;
use Gnikyt\BasicShopifyAPI\Options;
use Gnikyt\BasicShopifyAPI\Session;

// Create options for the API
$options = new Options();
$options->setVersion('2020-01');

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session('example.myshopify.com', 'access-token-here'));

// Now run your requests...
$result = $api->rest(...);
```

#### REST (async)

For REST calls, the shop domain and access token are required.

```php
use Gnikyt\BasicShopifyAPI\BasicShopifyAPI;
use Gnikyt\BasicShopifyAPI\Options;
use Gnikyt\BasicShopifyAPI\Session;

// Create options for the API
$options = new Options();
$options->setVersion('2020-01');

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session('example.myshopify.com', 'access-token-here'));

// Now run your requests...
$promise = $api->restAsync(...);
$promise->then(function (array $result) {
  // ...
});
```

#### GraphQL (sync)

For GraphQL calls, the shop domain and access token are required.

```php
use Gnikyt\BasicShopifyAPI\BasicShopifyAPI;
use Gnikyt\BasicShopifyAPI\Options;
use Gnikyt\BasicShopifyAPI\Session;

// Create options for the API
$options = new Options();
$options->setVersion('2020-01');

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session('example.myshopify.com', 'access-token-here'));

// Now run your requests...
$result = $api->graph(...);
```

#### GraphQL (async)

For GraphQL calls, the shop domain and access token are required.

```php
use Gnikyt\BasicShopifyAPI\BasicShopifyAPI;
use Gnikyt\BasicShopifyAPI\Options;
use Gnikyt\BasicShopifyAPI\Session;

// Create options for the API
$options = new Options();
$options->setVersion('2020-01');

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session('example.myshopify.com', 'access-token-here'));

// Now run your requests...
$promise = $api->graphAsync(...);
$promise->then(function (array $result) {
  // ...
});
```

#### Getting access (offline)

This is the default mode which returns a permanent token.
 
After obtaining the user's shop domain, to then direct them to the auth screen use `getAuthUrl`, as example (basic PHP):

```php
// Create options for the API
$options = new Options();
$options->setVersion('2020-01');
$options->setApiKey(env('SHOPIFY_API_KEY'));
$options->setApiSecret(env('SHOPIFY_API_SECRET'));

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session($_SESSION['shop']));

$code = $_GET['code'];
if (!$code) {
  /**
   * No code, send user to authorize screen
   * Pass your scopes as an array for the first argument
   * Pass your redirect URI as the second argument
   */
  $redirect = $api->getAuthUrl(env('SHOPIFY_API_SCOPES'), env('SHOPIFY_API_REDIRECT_URI'));
  header("Location: {$redirect}");
  exit;
} else {
  // We now have a code, lets grab the access token
  $api->requestAndSetAccess($code);

  // You can now make API calls
  $request = $api->rest('GET', '/admin/shop.json'); // or GraphQL
}
```

#### Getting access (per-user)

You can also change the grant mode to be `per-user` as [outlined in Shopify documentation](https://help.shopify.com/en/api/getting-started/authentication/oauth/api-access-modes). This will receieve user info from the user of the app within the Shopify store. The token recieved will expire at a specific time.

```php
// Create options for the API
$options = new Options();
$options->setVersion('2020-01');
$options->setApiKey(env('SHOPIFY_API_KEY'));
$options->setApiSecret(env('SHOPIFY_API_SECRET'));

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session($_SESSION['shop']));

$code = $_GET['code'];
if (!$code) {
  /**
   * No code, send user to authorize screen
   * Pass your scopes as an array for the first argument
   * Pass your redirect URI as the second argument
   * Pass your grant mode as the third argument
   */
  $redirect = $api->getAuthUrl(env('SHOPIFY_API_SCOPES'), env('SHOPIFY_API_REDIRECT_URI'), 'per-user');
  header("Location: {$redirect}");
  exit;
} else {
  // We now have a code, lets grab the access object
  $api->requestAndSetAccess($code);

  // You can now make API calls
  $request = $api->rest('GET', '/admin/shop.json'); // or GraphQL
}
```

#### Verifying HMAC signature

Simply pass in an array of GET params.

```php
// Will return true or false if HMAC signature is good.
$valid = $api->verifyRequest($_GET);
```

### Private API

This assumes you properly have your app setup in the partner's dashboard with the correct keys and redirect URIs.

#### REST

For REST (sync) calls, shop domain, API key, and API password are request

```php
// Create options for the API
$options = new Options();
$options->setType(true); // Makes it private
$options->setVersion('2020-01');
$options->setApiKey(env('SHOPIFY_API_KEY'));
$options->setApiPassword(env('SHOPIFY_API_PASSWORD'));

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session($_SESSION['shop']));

// Now run your requests...
$result = $api->rest(...);
```

#### GraphQL

For GraphQL calls, shop domain and API password are required.

```php
// Create options for the API
$options = new Options();
$options->setType(true); // Makes it private
$options->setVersion('2020-01');
$options->setApiPassword(env('SHOPIFY_API_PASSWORD'));

// Create the client and session
$api = new BasicShopifyAPI($options);
$api->setSession(new Session($_SESSION['shop']));

// Now run your requests...
$result = $api->graph(...);
```

### Making requests

#### REST

Requests are made using Guzzle.

```php
$api->rest(string $type, string $path, array $params = null, array $headers = [], bool $sync = true);
// or $api->getRestClient()->request(....);
```

+ `type` refers to GET, POST, PUT, DELETE, etc
+ `path` refers to the API path, example: `/admin/products/1920902.json`
+ `params` refers to an array of params you wish to pass to the path, examples: `['handle' => 'cool-coat']`
+ `headers` refers to an array of custom headers you would like to optionally send with the request, example: `['X-Shopify-Test' => '123']`
+ `sync` refers to if the request should be synchronous or asynchronous.

You can use the alias `restAsync` to skip setting `sync` to `false`.

##### If sync is true (regular rest call):

The return value for the request will be an array containing:

+ `response` the full Guzzle response object
+ `body` the JSON decoded response body (\Gnikyt\BasicShopifyAPI\ResponseAccess instance)
+ `errors` if any errors are detected, true/false
+ `exception` if errors are true, exception object is available
+ `status` the HTTP status code
+ `link` an array of previous/next pagination values, if available

*Note*: `request()` will alias to `rest()` as well.

##### If sync is false (restAsync call):

The return value for the request will be a Guzzle promise which you can handle on your own.

The return value for the promise will be an object containing:

+ `response` the full Guzzle response object
+ `body` the JSON decoded response body (\Gnikyt\BasicShopifyAPI\ResponseAccess instance)
+ `errors` if any errors are detected, true/false
+ `exception` if errors are true, exception object is available
+ `status` the HTTP status code
+ `link` an array of previous/next pagination values, if available

```php
$promise = $api->restAsync(...);
$promise->then(function (array $result) {
  // `response` and `body`, etc are available in `$result`.
});
```

##### Overriding request type

If you require the need to force a query string for example on a non-GET endpoint, you can specify the type as a key.

```php
$api->rest('PUT', '/admin/themes/12345/assets.json', ['query' => [...]]);
```

Valid keys are `query` and `json`.

##### Passing additional request options

If you'd like to pass additional request options to the Guzzle client created, pass them as the second argument of the constructor.

```php
// Create options for the API
$options = new Options();
$options->setVersion('2020-01');
// ...
$options->setGuzzleOptions(['connect_timeout' => 3.0]);

// Create the client
$api = new BasicShopifyApi($options);
```

#### GraphQL

Requests are made using Guzzle.

```php
$api->graph(string $query, array $variables = []);
```

+ `query` refers to the full GraphQL query
+ `variables` refers to the variables used for the query (if any)

The return value for the request will be an object containing:

+ `response` the full Guzzle response object
+ `body` the JSON decoded response body (\Gnikyt\BasicShopifyAPI\ResponseAccess instance)
+ `errors` if there was errors or not, will return the errors if any are found
+ `status` the HTTP status code

Example query:

```php
$result = $api->graph('{ shop { product(first: 1) { edges { node { handle, id } } } } }');
echo $result['body']['shop']['products']['edges'][0]['node']['handle']; // test-product
// or echo $result['body']->shop->products->edges[0]->node->handle;
```

Example mutation:

```php
$result = $api->graph(
    'mutation collectionCreate($input: CollectionInput!) { collectionCreate(input: $input) { userErrors { field message } collection { id } } }',
    ['input' => ['title' => 'Test Collection']]
);
echo $result['body']['collectionCreate']['collection']['id']; // gid://shopify/Collection/63171592234
// or echo $result['body']->collectionCreate->collection->id;
```

### API Versioning

This library supports [versioning the requests](https://www.shopify.com/partners/blog/api-versioning-at-shopify), example:

```php
// Create options for the API
$options = new Options();
$options->setVersion('2020-01'); // YYYY-MM or "unstable" is accepted

// Create the client
$api = new BasicShopifyAPI($options);
```

You can override the versioning at anytime for specific API requests, example:

```php
// Create options for the API
$options = new Options();
$options->setVersion('2020-01');

// Create the client
$api = new BasicShopifyAPI($options);
$api->rest('GET', '/admin/api/unstable/shop.json'); // Will ignore "2020-01" version and use "unstable" for this request
```

### Rate limiting

This library comes with a built-in basic rate limiter which utilizes `usleep` between applicable calls.

* For REST: it ensures you do not request more than the default of 2 calls per second.
* For GraphQL: it ensures you do not use more than the default of 50 points per second.

To adjust the default limits, use the option class' `setRestLimit` and `setGraphLimit`.

#### Custom rate limiting

You simply need to disable the built-in rate limiter and push in a custom Guzzle middleware. Example:

```php
$options = new Options();
// ...
$options->disableRateLimiting();

// ...
$api = new BasicShopifyAPI($options);
$api->addMiddleware(new CustomRateLimiter($api), 'rate:limiting');
```

### page_info / pagination support

2019-07 API version introduced a new `Link` header which is used for pagination ([explained here](https://help.shopify.com/en/api/guides/paginated-rest-results)).

If an endpoint supports page_info, you can use `$response->link` to grab the page_info value to pass in your next request.

Example:

```php
$response = $api->rest('GET', '/admin/products.json', ['limit' => 5]);
$link = $response['link']['next']; // eyJsYXN0X2lkIjo0MDkw
$link2 = $response['link']['previous']; // dkUIsk00wlskWKl
$response = $api->rest('GET', '/admin/products.json', ['limit' => 5, 'page_info' => $link]);
```

### Isolated API calls

You can initialize the API once and use it for multiple shops. Each instance will be contained to not pollute the others. This is useful for something like background job processing.

```php
$api->withSession(Session $newSession, Closure $closure);
```

`$this` will be binded to the closure. Example:

```php
$api->withSession(new Session('someshop.myshopify.com', 'some-token'), function (): void {
  $request = $this->rest('GET', '/admin/shop.json');
  echo $request['body']['shop']['name']; // Some Shop
});

// $api->rest/graph will not be affected by the above code, it will use previously defined session
```

### Retries

This library utilizes `caseyamcl/guzzle_retry_middleware` middleware package.

By default, `429`, '500` and `503` errors will be retried twice.

For REST calls, it will utilize Shopify's `X-Retry-After` header to wait *x* seconds before retrying the call.

When all retries are exhasted, the standard response from the library will return where you can handle the error.

To change the status codes watched or the maximum number of retries, use the option class' `setGuzzleOptions`:

```php
// Create options for the API
$options = new Options();
$options->setVersion('2020-01');
// ...
$options->setGuzzleOptions(
  'max_retry_attempts' => 3, // Was 2
  'retry_on_status'    => [429, 503, 400], // Was 439, 503, 500
);

// Create the client
$api = new BasicShopifyApi($options);
```

### Errors

This library internally catches only 400-500 status range errors through Guzzle. You're able to check for an error of this type and get its response status code and body.

```php
$call = $api->rest('GET', '/admin/non-existant-route-or-object.json');

if ($call['errors']) {
  echo "Oops! {$call['status']} error";
  var_dump($call['body']);

  // Original exception can be accessed via `$call['exception']`
  // Example, if response body was `{"error": "Not found"}`...
  /// then: `$call['body']` would return "Not Found"
}
```

### Middleware

This library takes advantage of using Guzzle middleware for request/response checks and modifications. You're also able to inject middleware.

```php
$api->addMiddleware([callable]);
```

See Guzzle's documentation on middleware. As well, you can browse this library's middleware for examples.

### Storage

For storing the current request times, API limits, request costs, etc. A basic in-memory array store is used `Gnikyt\BasicShopifyAPI\Store\Memory`.

If you would like to implement a more advananced store such as one with Redis, simply implement `Gnikyt\BasicShopifyAPI\Contracts\StateStorage` and set the client to use it, example:

```php
$timeStore = new RedisStore();
$limitStore = new RedisStore();

$api = new BasicShopifyAPI($options, $timeStore, $limitStore);
```

## Documentation

Code documentation is [available here](https://gnikyt.com/Basic-Shopify-API) from phpDocumentor via `phpdoc -d src -t doc`.

## LICENSE

This project is released under the MIT [license](https://github.com/gnikyt/Basic-Shopify-API/blob/master/LICENSE).

## Misc

Using Python? [Check out basic_shopify_api](https://github.com/gnikyt/basic_shopify_api).


CHANGELOG.md
# CHANGELOG

## 11.0.0

* Changes namespace to Gnikyt

## 10.0.6

* Changes return type of iterable to match signature of iterable for PHP 8.1 (#129)

## 10.0.5

* Bugfix to rate limiter (#123)

## 10.0.4

* Bugfix to rate limiter (#120)

## 10.0.3

* Moved `final` flag to the constructor

## 10.0.2

* Added `final` flag to `ResponseAccess` class
* CI improvements

## 10.0.1

* Fix for Graph API limit calls not accessing the correct index for last call data
* Update for microtime for rate limiter

## 10.0.0

* Bump Guzzle to version 8 from 7 (#83)

## 9.1.4

* Implemented toArray method, mainly used for Laravel integration

## 9.1.3

* Fix to allow for overriding request type for data (#80)

## 9.1.2

* Oauth error handling (#77).

## 9.1.1

* Implemented `Iterable` and `Countable` on `ResponseAccess`
* Added methods `keys` and `values` to allow for getting keys and values easier
* Add JSON serilize method

## 9.1.0

* Rate limiting for REST has been improved to use a moving window method
* `Sleep` deferrer now returns a float for the time microtime
* `Memory` store now has a `reset` method to support the new rate limiting
* `Memory` now accepts more then two values to support the new rate limiting

## 9.0.0

*Contains breaking changes*; Consult `UPGRADING.md`.

* Library refactored and split up, no longer one file
* Auth header handling moved into Guzzle middleware
* API versioning handler moved into Guzzle middleware
* Rate limiting improved and now available for GraphQL
* Options moved into its own configuration class
* "Retry" ability now built-in via `guzzle_retry_middleware` which respects Shopify's 'X-Retry-After' header
* API responses are now mapped into a response class which can be accessed as an array or object
* Storage of how/where rate limits, request times, and others are now changable with the `StateStorage` interface, and a default in-memory storage class provided
* Both GraphQL and REST success and error returns have been normalized

## 8.2.0

* Added `bodyArray` to response object of rest and graph, this is the same as `body` except instead of a stdClass being returned, its an array.

## 8.1.0

* Added response to error callback function (#64)
* Allow for additional client creation options (#65)

## 8.0.0

* Added return types to all functions.
* Added async support for GraphQL via `graphAsync`.
* Added support for call to `/admin/oauth/access_scopes.json`.

## 6.1.2

* Fix for REST calls not properly building error object (#47).

## 6.1.1

* Fixed `extractHeaderLink` for #45, to include both `next` and `previous`.

## 6.1.0

* Adds support for `Link` header (#39)

## 6.0.0

* Added ability to do Guzzle sync and async requests through Promises.
* `errors` now returns a boolean instead of an object. `body` now contains the error response.

## 5.5.0

* Added ability to use custom headers in REST calls.

## 5.4.0

* Added option to supply a PSR-compatible logger via `setLogger`.

## 5.3.3

* Added matching to ignore certain oauth paths.

## 5.3.2

* Added for for issue #24 in detection of what an authable request is.

## 5.3.1

* Fixed merged in to prevent versioning on access token calls.

## 5.3.0

* Added support for versioned API calls for both REST and GraphAPI
* New `setVersion(string $version)` method added, and new `getVersion()` method added

## 5.2.0

* Added ability for `per-user` authentication. There is now ability to set/grab a user from authentication.
* `requestAndSetAccess(string $code)` method added which will automatically set the access token and user (if applicable) to the API instance for you as a helper method.

## 5.1.0

* Added a `authRequest` middleware to Guzzle which handles adding all needed headers and checking required API keys to run those requests
* Fixed issue for redirect not working

## 5.0.0

*Possible breaking release depending on how you handle errors*

* 400-500 errors are now captured internally and accessible through the resulting object (#16)
* Middleware was added to the Guzzle requests to fix redirections (#16)

## 4.0.2

* Changes to the response of GraphQL calls to better check for errors.

## 4.0.1

* Update for more accurate timing

## 4.0.0

* Added rate limiting abilities (basic)

## 3.0.3

* Fix for #13 for requests where call limit header is not always supplied

## 3.0.2

* Adjusted API to work better with Shopify's implementation of GraphQL (#10)
* `graph()` call now accepts two arguments, `graph(string $query, array $variables = [])`

## 3.0.1

* Fix to obtaining access token

## 3.0.0

*Contains breaking changes*

To better the library, it has been reverted back to its original single-class form and backwards compatibile with 1.x.x

* GraphQL and REST are all under one class
* `getApiCalls()` now takes two arguments, first being rest|graph, second being the key
* `rest()` is now for REST calls
* `graph()` is now for GraphQL calls
* `request()` is aliased to `rest()` for backward compatibility

## 2.0.0

*Contains breaking changes*

* No longer a single file, it now namespaced under `OhmyBrew\ShopifyAPI`
* GraphQL is now introduced under `Gnikyt\ShopifyAPI\GraphAPI`
* REST is moved to `Gnikyt\ShopifyAPI\RestAPI`

## 1.0.1

* Fixed issue #3 by @ncpope, for newly created stores (after December 2017) not accepting GET body

## 1.0.0

* Basic API code implemented
* Tests completed with full coverage


_test/GraphApiTest.php
<?php

namespace Osiset\Test;

use Exception;
use GuzzleHttp\Psr7\Response;
use Osiset\BasicShopifyAPI;

class GraphApiTest extends BaseTest
{
    /**
     * Setup for phpUnit.
     *
     * @return void
     */
    public function setUp()
    {
        // Query call
        $this->query = [
            '{ shop { products(first: 1) { edges { node { handle id } } } } }',
        ];

        // Mutation call with variables
        $this->mutation = [
            'mutation collectionCreate($input: CollectionInput!) { collectionCreate(input: $input) { userErrors { field message } collection { id } } }',
            ['input' => ['title' => 'Test Collection']],
        ];
    }

    /**
     * @test
     *
     * Checking base URL
     */
    public function itShouldReturnBaseUrl()
    {
        $responses = [
            new Response(
                200,
                [],
                file_get_contents(__DIR__.'/fixtures/graphql/shop_products.json')
            ),
        ];

        $api = new BasicShopifyAPI(true);
        $mock = $this->buildClient($api, $responses);

        $api->setShop('example.myshopify.com');
        $api->setApiKey('123');
        $api->setApiPassword('abc');
        $api->graph($this->query[0]);

        $lastRequest = $mock->getLastRequest()->getUri();
        $this->assertEquals('https', $lastRequest->getScheme());
        $this->assertEquals('example.myshopify.com', $lastRequest->getHost());
        $this->assertEquals('/admin/api/graphql.json', $lastRequest->getPath());
    }

    /**
     * @test
     * @expectedException Exception
     * @expectedExceptionMessage Shopify domain missing for API calls
     *
     * Ensure Shopify domain is there for queries
     */
    public function itShouldThrowExceptionForMissingDomainOnQuery()
    {
        $api = new BasicShopifyAPI();
        $api->graph($this->query[0]);
    }

    /**
     * @test
     * @expectedException Exception
     * @expectedExceptionMessage API password/access token required for private Shopify GraphQL calls
     *
     * Ensure API password is there for private queries
     */
    public function itShouldThrowExceptionForMissingApiPasswordOnPrivateQuery()
    {
        $api = new BasicShopifyAPI(true);
        $api->setShop('example.myshopify.com');
        $api->graph($this->query[0]);
    }

    /**
     * @test
     * @expectedException Exception
     * @expectedExceptionMessage Access token required for public Shopify GraphQL calls
     *
     * Ensure access token is there for public queries
     */
    public function itShouldThrowExceptionForMissingAccessTokenOnPublicQuery()
    {
        $api = new BasicShopifyAPI();
        $api->setShop('example.myshopify.com');
        $api->graph($this->query[0]);
    }

    /**
     * @test
     *
     * Should get Guzzle response and JSON body for success
     */
    public function itShouldReturnGuzzleResponseAndJsonBodyForSuccess()
    {
        $responses = [
            new Response(
                200,
                [],
                file_get_contents(__DIR__.'/fixtures/graphql/shop_products.json')
            ),
        ];

        $api = new BasicShopifyAPI();
        $mock = $this->buildClient($api, $responses);

        $api->setShop('example.myshopify.com');
        $api->setAccessToken('!@#');

        // Fake param just to test it receives it
        $request = $api->graph($this->query[0]);
        $data = $mock->getLastRequest()->getUri()->getQuery();
        $token_header = $mock->getLastRequest()->getHeader('X-Shopify-Access-Token')[0];

        // Assert the response data
        $this->assertEquals(true, is_object($request));
        $this->assertInstanceOf('GuzzleHttp\Psr7\Response', $request->response);
        $this->assertEquals(200, $request->response->getStatusCode());
        $this->assertEquals(false, $request->errors);
        $this->assertEquals(true, is_object($request->body));
        $this->assertEquals(true, is_array($request->bodyArray));
        $this->assertEquals('gift-card', $request->body->shop->products->edges[0]->node->handle);
        $this->assertEquals('gift-card', $request->bodyArray['shop']['products']['edges'][0]['node']['handle']);
        $this->assertEquals('!@#', $token_header);

        // Confirm limits have been updated
        $this->assertEquals(5, $api->getApiCalls('graph', 'made'));
        $this->assertEquals(1000, $api->getApiCalls('graph', 'limit'));
        $this->assertEquals(1000 - 5, $api->getApiCalls('graph', 'left'));
        $this->assertEquals(['left' => 1000 - 5, 'made' => 5, 'limit' => 1000, 'requestedCost' => 5, 'actualCost' => 5, 'restoreRate' => 50], $api->getApiCalls('graph'));
    }

    /**
     * @test
     *
     * Should get Guzzle response and JSON body for error
     */
    public function itShouldReturnGuzzleResponseForError()
    {
        $responses = [
            new Response(
                200,
                [],
                file_get_contents(__DIR__.'/fixtures/graphql/shop_products_error.json')
            ),
        ];

        $api = new BasicShopifyAPI();
        $mock = $this->buildClient($api, $responses);

        $api->setShop('example.myshopify.com');
        $api->setAccessToken('!@#');

        // Fake param just to test it receives it
        $request = $api->graph($this->query[0]);
        $data = $mock->getLastRequest()->getUri()->getQuery();
        $token_header = $mock->getLastRequest()->getHeader('X-Shopify-Access-Token')[0];

        // Assert the response
        $this->assertEquals(true, is_object($request));
        $this->assertInstanceOf('GuzzleHttp\Psr7\Response', $request->response);
        $this->assertEquals(200, $request->response->getStatusCode());
        $this->assertEquals(true, $request->errors);
        $this->assertEquals(true, is_array($request->body));
        $this->assertEquals("Field 'productz' doesn't exist on type 'Shop'", $request->body[0]->message);
        $this->assertEquals('!@#', $token_header);

        // Confirm limits have not been updated since there is no cost
        $this->assertEquals(0, $api->getApiCalls('graph', 'made'));
        $this->assertEquals(1000, $api->getApiCalls('graph', 'limit'));
        $this->assertEquals(0, $api->getApiCalls('graph', 'left'));
        $this->assertEquals(['left' => 0, 'made' => 0, 'limit' => 1000, 'restoreRate' => 50, 'requestedCost' => 0, 'actualCost' => 0], $api->getApiCalls('graph'));
    }

    /**
     * @test
     *
     * Ensure Async requests are working
     */
    public function itShouldRunAsyncRequests()
    {
        $responses = [
            new Response(
                200,
                [],
                file_get_contents(__DIR__.'/fixtures/graphql/shop_products.json')
            ),
        ];

        $api = new BasicShopifyAPI();
        $this->buildClient($api, $responses);

        $api->setShop('example.myshopify.com');
        $api->setAccessToken('!@#');

        $promise = $api->graphAsync($this->query[0]);
        $promise->then(function ($result) {
            $this->assertEquals(true, is_object($result->body));
        });
        $promise->wait();
    }

    /**
     * @test
     *
     * Should process query with variables
     */
    public function itShouldProcessQueryWithVariables()
    {
        $responses = [
            new Response(
                200,
                [],
                file_get_contents(__DIR__.'/fixtures/graphql/create_collection.json')
            ),
        ];

        $api = new BasicShopifyAPI();
        $mock = $this->buildClient($api, $responses);

        $api->setShop('example.myshopify.com');
        $api->setAccessToken('!@#');

        // Fake param just to test it receives it
        $request = $api->graph($this->mutation[0], $this->mutation[1]);
        $data = $mock->getLastRequest()->getUri()->getQuery();
        $token_header = $mock->getLastRequest()->getHeader('X-Shopify-Access-Token')[0];

        // Assert the response data
        $this->assertEquals(true, is_object($request));
        $this->assertInstanceOf('GuzzleHttp\Psr7\Response', $request->response);
        $this->assertEquals(200, $request->response->getStatusCode());
        $this->assertEquals(true, is_object($request->body));
        $this->assertEquals('gid://shopify/Collection/63171592234', $request->body->collectionCreate->collection->id);
        $this->assertEquals('!@#', $token_header);

        // Confirm limits have been updated
        $this->assertEquals(11, $api->getApiCalls('graph', 'made'));
        $this->assertEquals(1000, $api->getApiCalls('graph', 'limit'));
        $this->assertEquals(1000 - 11, $api->getApiCalls('graph', 'left'));
        $this->assertEquals(['left' => 1000 - 11, 'made' => 11, 'limit' => 1000, 'requestedCost' => 11, 'actualCost' => 11, 'restoreRate' => 50], $api->getApiCalls('graph'));
    }

    /**
     * @test
     * @expectedException Exception
     * @expectedExceptionMessage Invalid API call limit key. Valid keys are: left, made, limit, restoreRate, requestedCost, actualCost
     *
     * Ensure we pass a valid key to the API calls
     */
    public function itShouldThrowExceptionForInvalidApiCallsKey()
    {
        $api = new BasicShopifyAPI();
        $api->getApiCalls('graph', 'oops');
    }

    /**
     * @test
     *
     * Should version API paths if a version is set.
     */
    public function itShouldVersionApiPaths()
    {
        $responses = [];
        for ($i = 0; $i < 2; $i++) {
            $responses[] = new Response(
                200,
                [],
                file_get_contents(__DIR__.'/fixtures/graphql/shop_products.json')
            );
        }

        $api = new BasicShopifyAPI(true);
        $api->setShop('example.myshopify.com');
        $api->setApiKey('123');
        $api->setApiPassword('abc');
        $mock = $this->buildClient($api, $responses);

        // No version set
        $api->graph($this->query[0]);
        $lastRequest = $mock->getLastRequest()->getUri();
        $this->assertEquals('/admin/api/graphql.json', $lastRequest->getPath());

        // A set version
        $api->setVersion('2020-01');
        $api->graph($this->query[0]);
        $lastRequest = $mock->getLastRequest()->getUri();
        $this->assertEquals('/admin/api/2020-01/graphql.json', $lastRequest->getPath());
    }
}
